### PR TITLE
Add support for Login with Multi-Factor Authentication

### LINKED ISSUE #726: Login to support multi-factor authentication
**Describe the feature**  
Login to support supplying otp key.
Login method to have otp key parameter and must be able to supply the otp key during login.

**Explain why this feature is needed**  
D365 users (in our case) will be asked always for an otp key since MFA is a security feature for Dynamics 365. The key will be coming from Microsoft Authenticator App 
Untitled
(instead of sms).

**Additional context**

### PR DESCRIPTION
**Pull request by AngelRodriguez8008**
### Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (updates to documentation, formatting, etc.)

### Description
Add support for Login with Multi-Factor Authentication
This [comment](https://github.com/microsoft/EasyRepro/issues/726#issuecomment-577711262) documents the configuration, consider to add  to the project Wiki

### Issues addressed
Fix #726 Login to support multi-factor authentication (related to  #653)
Fix SeleniumExtensions.WaitUntilAvailable (see my comment in #707)
May fix: #721

### All submissions:
- [x] My code follows the code style of this project.
- [x] Do existing samples that are effected by this change still run?
- [x] I have added samples for new functionality. 
- [x] I raise detailed error messages when possible.
- [x] My code does not rely on labels that have the option to be hidden.

### Which browsers was this tested on?
- [x] Chrome
- [ ] Firefox
- [ ] IE
- [ ] Edge

@dtu11 it was hard to merge with #734 can you please take a look at the methods:
- WebClient.OpenApp:331: OpenApp
- WebClient.Login:118: Login

@TYLEROL in commit: [Add MFA Key to UCI Tests](https://github.com/microsoft/EasyRepro/commit/fbfc3c758b12ef2bef0ee80f4ba2f6ee918d4b40), I add the MFA SecrectKey to all UCI Tests. It is a simple change, however, I have none Environment without MFA to tests what happen if is MFA is not required & this Key is set to empty in app.config. Can you please check if it is working? 
I wanted add the TestsBase to all UCI sample methods. I stops to doing it because for some unknown reason it broke the UCI Ordered Tests file and I was not sure if it was really desired; for samples porpoise may be not good.

### CODE REVIEW

File: Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs [hunk 1]
```diff
@@ -547,74 +555,56 @@ public static object WaitForScript(this IWebDriver driver, string script, TimeSp
...
+                    ? elements.FirstOrDefault(x => x?.Displayed == true)
+                    : elements.First(x => x != null);
+                
+                return result;
+            };
+            return WaitUntil(driver, conditions, timeout ?? Constants.DefaultTimeout, successCallback, failureCallback);
         }
 
-        public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, TimeSpan timeout, Action<IWebDriver> successCallback, Action<IWebDriver> failureCallback)
+        private static IWebElement WaitUntil(IWebDriver driver, Func<IWebDriver, IWebElement> conditions, TimeSpan? timeout, Action<IWebDriver> successCallback, Action<IWebDriver> failureCallback)
```
Comment (easyrepro) [inline Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs]: can you make it public. we can use it to implement WaitForNotXXXX when specific methods are required

Comment (AngelRodriguez8008) [reply Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs:587]: thanks. This PR is big enough, I plan to change/extend the Selenium Extensions in other PR. In this case I was just fixing a bug.

**Review by AngelRodriguez8008 — COMMENTED** [review]
Please @TYLEROL approve this PR

File: Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs (hunk 1 above)
Comment (AngelRodriguez8008) [reply Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs:587]: Hi @easyrepro. You can check the SeleniumExtensions.cs in #747 the Wait-Methods are now extended. There is also two WaitUntil that you can extend as you want & also a RepeatUntil. I hope did you like it. BR

**Review by TYLEROL — APPROVED** [review]

**Comment by AngelRodriguez8008** [comment]
Thank you, very much @TYLEROL

**Comment by tristanperper** [comment]
Hi @AngelRodriguez8008 ,

I see the fix has been merged into develop. Should the Login With MFA be included into a new release anytime soon?

**Comment by AngelRodriguez8008** [comment]
Hi @tristanperper,

Sorry I guest so. But I don't manage the releases, I also work  just with the version in 'develop' branch or or one version of my fork. I don't know how oft will be or has been released.

br

**Pull request merged**